<style>
/* Main Title */
h1 {
    color: #1E293B;
    text-align: center;
    font-weight: 800;
    margin-top: 20px;
}

/* Subtitle */
h3 {
    text-align: center;
    color: #2F855A;
    font-weight: 500;
}

/* Section Headers */
h2 {
    color: #2F855A;
    border-bottom: 3px solid #2F855A;
    padding-bottom: 6px;
    margin-top: 50px;
}

/* Insight Box */
.insight-box {
    background-color: #F6E7D8;
    padding: 18px;
    border-left: 6px solid #2F855A;
    border-radius: 8px;
    margin: 20px 0;
    font-size: 15px;
}

/* Alert Box */
.alert-box {
    background-color: #FFF5F0;
    padding: 18px;
    border-left: 6px solid #C05621;
    border-radius: 8px;
    margin: 20px 0;
    font-weight: 500;
}

/* Dashboard Cards */
.metric-card {
    background-color: #F8FAFC;
    padding: 20px;
    border-radius: 10px;
    box-shadow: 0px 4px 12px rgba(0,0,0,0.05);
    width: 240px;
}
</style>


<div style="text-align:center; padding:10px;">
    <h1><b>Predicting County-Level Food Insecurity in Kenya Using Machine Learning</b></h1>
    <h3>Early Warning System for Acute Food Crisis</h3>
    <p><strong> Capstone Project</strong></p>
    <p style="color:#334155; font-size:14px; margin-top:5px;">👤Author: Bleezy</p>
</div>
<hr style="height:3px;background-color:#2F855A;border:none;">


<h2><span style=<b>📋 Table of Contents</b></span></h2>

<details>
<summary style="cursor:pointer; font-size:16px; font-weight:bold; color:#2E86C1; padding:6px 0;">1. Business Understanding</summary>
<ul>
  <li><a href="#11-background">1.1 Background</a></li>
  <li><a href="#12-the-problem">1.2 The Problem</a></li>
  <li><a href="#13-project-goals">1.3 Project Goals</a></li>
  <li><a href="#14-stakeholders--users">1.4 Stakeholders & Users</a></li>
  <li><a href="#15-measuring-success">1.5 Measuring Success</a></li>
  <li><a href="#16-expected-benefits">1.6 Expected Benefits</a></li>
</ul>
</details>

<details>
<summary style="cursor:pointer; font-size:16px; font-weight:bold; color:#2E86C1; padding:6px 0;">2. Data Understanding</summary>
<ul>
  <li><a href="#21-import-libraries">2.1 Import Libraries</a></li>
  <li><a href="#22-load-data">2.2 Load Data</a></li>
  <li><a href="#23-initial-data-exploration">2.3 Initial Data Exploration</a></li>
  <li><a href="#24-data-quality-check">2.4 Data Quality Check</a></li>
  <li><a href="#25-basic-statistics">2.5 Basic Statistics</a></li>
  <li><a href="#26-unique-values-in-key-columns">2.6 Unique Values in Key Columns</a></li>
  <li><a href="#27-date-ranges-for-each-dataset">2.7 Date Ranges for Each Dataset</a></li>
  <li><a href="#28-exploratory-data-analysis">2.8 Exploratory Data Analysis</a></li>
</ul>
</details>

<details>
<summary style="cursor:pointer; font-size:16px; font-weight:bold; color:#2E86C1; padding:6px 0;">3. Data Preprocessing</summary>
<ul>
  <li><a href="#31-data-cleaning">3.1 Data Cleaning</a></li>
  <li><a href="#32-post-cleaning-exploratory-analysis">3.2 Post-Cleaning Exploratory Analysis</a></li>
  <li><a href="#33-feature-engineering">3.3 Feature Engineering</a></li>
  <li><a href="#34-data-merging">3.4 Data Merging</a></li>
  <li><a href="#35-multicollinearity-check">3.5 Multicollinearity Check</a></li>
  <li><a href="#36-modeling-preprocessing">3.6 Modeling Preprocessing</a></li>
</ul>
</details>

<details>
<summary style="cursor:pointer; font-size:16px; font-weight:bold; color:#2E86C1; padding:6px 0;">4. Modeling</summary>
<ul>
  <li><a href="#41-baseline-model--logistic-regression">4.1 Baseline Model — Logistic Regression</a></li>
  <li><a href="#42-model-performance-comparison">4.2 Model Performance Comparison</a></li>
  <li><a href="#43-hyperparameter-tuning">4.3 Hyperparameter Tuning</a></li>
  <li><a href="#44-validation-on-held-out-test-set">4.4 Validation on Held-Out Test Set</a></li>
  <li><a href="#45-threshold-optimization">4.5 Threshold Optimization</a></li>
  <li><a href="#46-feature-importance-analysis">4.6 Feature Importance Analysis</a></li>
  <li><a href="#47-simplified-model">4.7 Simplified Model</a></li>
  <li><a href="#48-saving-the-final-model">4.8 Saving the Final Model</a></li>
</ul>
</details>

<details>
<summary style="cursor:pointer; font-size:16px; font-weight:bold; color:#2E86C1; padding:6px 0;">5. Evaluation</summary>
<ul>
  <li><a href="#51-confusion-matrix--roc-curve">5.1 Confusion Matrix & ROC Curve</a></li>
  <li><a href="#52-precision-recall-curve--feature-importance">5.2 Precision-Recall Curve & Feature Importance</a></li>
  <li><a href="#53-county-prediction-map">5.3 County Prediction Map</a></li>
  <li><a href="#54-model-comparison-summary">5.4 Model Comparison Summary</a></li>
  <li><a href="#55-threshold-analysis">5.5 Threshold Analysis</a></li>
  <li><a href="#56-temporal-prediction-analysis">5.6 Temporal Prediction Analysis</a></li>
  <li><a href="#57-county-risk-profile">5.7 County Risk Profile</a></li>
  <li><a href="#58-final-model-report">5.8 Final Model Report</a></li>
  <li><a href="#59-shap-analysis--model-interpretability">5.9 SHAP Analysis — Model Interpretability</a></li>
</ul>
</details>

<details>
<summary style="cursor:pointer; font-size:16px; font-weight:bold; color:#2E86C1; padding:6px 0;">6. Deployment</summary>
<ul>
  <li><a href="#61-prediction-pipeline">6.1 Prediction Pipeline</a></li>
  <li><a href="#62-pipeline-validation">6.2 Pipeline Validation</a></li>
  <li><a href="#63-saving-pipeline-outputs">6.3 Saving Pipeline Outputs</a></li>
  <li><a href="#64-limitations--future-work">6.4 Limitations & Future Work</a></li>
</ul>
</details>

<details>
<summary style="cursor:pointer; font-size:16px; font-weight:bold; color:#2E86C1; padding:6px 0;">7. Conclusion</summary>
<ul>
  <li><a href="#71-project-summary">7.1 Project Summary</a></li>
  <li><a href="#72-key-findings">7.2 Key Findings</a></li>
  <li><a href="#73-business-goals-vs-outcomes">7.3 Business Goals vs Outcomes</a></li>
  <li><a href="#74-reflections">7.4 Reflections</a></li>
</ul>
</details>

</div>

## <span style="color:#2E86C1;"><b> 1. Business Understanding </b></span>

### **1.1 Background**

Food shortages remain a <span style="color:#C0392B;"><b>serious and persistent challenge</b></span> in Kenya.  
Each year, millions of people struggle to access sufficient and nutritious food.

Approximately <span style="color:#8E44AD;"><b>80% of Kenya’s land</b></span> consists of 
<b>Arid and Semi-Arid Lands (ASALs)</b>, home to more than 
<span style="color:#8E44AD;"><b>10 million people</b></span>.

<span style="color:#D35400;">Key Characteristics of ASAL Areas</span>

- 🌧️ Low and unpredictable rainfall  
- 🌱 Poor agricultural conditions  
- 🐄 Heavy dependence on livestock  
- 🌵 High vulnerability to drought

The IPC system (Integrated Food Security Phase Classification) is used worldwide to 
measure food insecurity. It classifies food insecurity into <b>5 phases</b>: 

| Phase | Classification | Meaning |
|-------|---------------|----------|
| 1 | <span style="color:#27AE60;"><b>Minimal</b></span> | Food needs are met |
| 2 | <span style="color:#F39C12;"><b>Stressed</b></span> | Households struggle |
| 3 | <span style="color:#E67E22;"><b>Crisis</b></span> | Urgent assistance required |
| 4 | <span style="color:#E74C3C;"><b>Emergency</b></span> | Severe food gaps |
| 5 | <span style="color:#922B21;"><b>Famine</b></span> | Extreme starvation |


> <span style="color:#C0392B;"><b>Phase 3 or higher = Immediate humanitarian intervention required.</b></span>

In Kenya, IPC assessments are done twice a year by organizations like:

- NDMA  (National Drought Management Authority)
- FAO   (Food and Agriculture Organization)
- WFP   (World Food Programme)

These reports describe the <b>current situation</b> but do <span style="color:#C0392B;"><b>not predict future crises</b></span>. 
This means: 
-  Help often arrives late  
-  Resources are mobilized after crisis escalation  
-  Preventive action becomes difficult  

Building a predictive system could enable:

- Early intervention  
- Better resource allocation  
- Reduced humanitarian impact

### **1.2 The Problem**

Kenya lacks a strong predictive system for food crises at the county level.

### Because of this:

- Delayed humanitarian response  
- Inefficient use of resources  
- Limited preventive action  
- Increased human suffering  
- Higher emergency costs 

> 💡 A predictive system could improve planning, prioritization, and coordination.


### **1.3 Project Goals**

 **Main Goal**

Build a **machine learning model** that predicts which counties will reach  
<span style="color:#C0392B;"><b>IPC Phase 3 or higher</b></span> within the next **1–3 months**.

**Model Inputs**

- Rainfall data  
- Food price data  
- Past IPC classifications  


**Additional Goals**

- Identify the main drivers of food insecurity  
- Measure prediction accuracy using public data  
- Create a simple tool for non-technical planners  
- Develop a framework applicable to other East African countries  


### **1.4 Stakeholders & Users**

| Stakeholder | Need | How Predictions Help |
|-------------|------|---------------------|
| WFP & FAO | Plan food delivery | 2–3 months advance warning |
| County Governments | Request funding | Early emergency planning |
| NGOs | Allocate limited funds | Prioritize high-risk areas |
| NDMA | Coordinate drought response | Support national planning |
| Donors (USAID, etc.) | Fund programs | Evidence-based funding decisions |


### **1.5 Measuring Success**

**Technical Success**

- ≥ 80% of real crises correctly predicted (**High Recall**)  
- ≥ 70% of predicted crises actually occur (**Good Precision**)  
- ROC-AUC ≥ 0.80  
- Predictions made 1–3 months in advance  

**Business Success**

- Clear explanation of why a county is at risk  
- Easy for non-technical users  
- Uses public, regularly updated data  
- Aligns with real-world humanitarian understanding  

### **1.6 Expected Benefits**

**Cost Savings**

Early action significantly reduces costs.

***Savings Include:***

- Lower emergency procurement costs  
- Improved transport planning  
- Fewer livestock deaths  
- Reduced long-term poverty  

 **Human Impact**

- Prevent crisis escalation into famine  
- Reduce child malnutrition  
- Protect livelihoods  
- Minimize long-term economic damage  



**Operational Benefits**

- Stronger inter-agency coordination  
- Faster response times  
- Better funding justification  
- Shift from emergency response to resilience building  

## <span style="color:#2E86C1;"><b>2. Data Understanding<b></span>

### **2.1 Import Libraries**
<p>Before we begin, we load all the tools required for this project.</p>



In [8]:
# ==============================
# Data Manipulation & Analysis
# ==============================
import pandas as pd
import numpy as np

# ==============================
# Data Visualization
# ==============================
import matplotlib.pyplot as plt
import seaborn as sns

# ==============================
# Machine Learning — Preprocessing
# ==============================
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import (
    train_test_split,
    StratifiedKFold,
    cross_val_score,
    RandomizedSearchCV
)

# ==============================
# Machine Learning — Models
# ==============================
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

# ==============================
# Machine Learning — Metrics
# ==============================
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    ConfusionMatrixDisplay,
    RocCurveDisplay,
    PrecisionRecallDisplay,
    f1_score,
    precision_score,
    recall_score,
    roc_auc_score
)

# ==============================
# Imbalanced Data Handling
# ==============================
from imblearn.over_sampling import SMOTE

# ==============================
# Model Persistence & Utilities
# ==============================
import joblib
import json
import os

# ==============================
# Geospatial
# ==============================
import geopandas as gpd

# ==============================
# Model Interpretability
# ==============================
import shap

# ==============================
# Display Settings
# ==============================
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

sns.set_style("whitegrid")
sns.set_palette("muted")

print("✓ All libraries loaded successfully!")
print(f"   Pandas:  {pd.__version__}")
print(f"   Numpy:   {np.__version__}")

✓ All libraries loaded successfully!
   Pandas:  3.0.1
   Numpy:   1.26.4


### **2.2. Load Data**
<p>Load all five datasets into the notebook for analysis.</p>

<p><strong>Datasets Overview:</strong></p>
<ul>
<li><strong>IPC Data</strong> – Records of food insecurity levels (our target variable)</li>
<li><strong>Rainfall Data</strong> – Climate patterns and rainfall measurements</li>
<li><strong>Food Prices</strong> – Market-level prices indicating supply and accessibility</li>
<li><strong>Conflict Data</strong> – Geolocated records of violence and unrest</li>
<li><strong>MPI Data</strong> – Multidimensional Poverty Index (MPI) showing socioeconomic vulnerability</li>



In [10]:
# Set the path where your data files are located
data_path = '../data/raw/'

In [11]:
print("\033[1;34mLoading datasets...\033[0m\n")  

# 1. Load IPC Data (Food Insecurity)
print("\033[1;32m1. Loading IPC data...\033[0m") 
ipc_df = pd.read_csv(data_path + 'ipc_ken_area_long.csv')
print(f"   ✓ Loaded {len(ipc_df):,} rows")

# 2. Load Rainfall Data
print("\n\033[1;32m2. Loading Rainfall data...\033[0m")
rainfall_df = pd.read_csv(data_path + 'ken-rainfall-subnat-full.csv')
print(f"   ✓ Loaded {len(rainfall_df):,} rows")

# 3. Load Food Prices Data
print("\n\033[1;32m3. Loading Food Prices data...\033[0m")
prices_df = pd.read_csv(data_path + 'wfp_food_prices_ken.csv', skiprows=1)
print(f"   ✓ Loaded {len(prices_df):,} rows")

# 4. Load Conflict Data (Excel)
print("\n\033[1;32m4. Loading Conflict data...\033[0m")
conflict_df = pd.read_excel(
    data_path + 'kenya_political_violence_events_and_fatalities_by_month-year_as-of-11feb2026.xlsx',
    sheet_name='Data'
)
print(f"   ✓ Loaded {len(conflict_df):,} rows")

# 5. Load MPI Data (Poverty Index)
print("\n\033[1;32m5. Loading MPI (Poverty) data...\033[0m")
mpi_df = pd.read_csv(data_path + 'ken_mpi.csv', skiprows=1)
print(f"   ✓ Loaded {len(mpi_df):,} rows")

print("\033[1;34m✓ ALL DATASETS LOADED Sawa sawa!\033[0m")

Loading datasets...

1. Loading IPC data...
   ✓ Loaded 4,522 rows

2. Loading Rainfall data...
   ✓ Loaded 131,544 rows

3. Loading Food Prices data...
   ✓ Loaded 17,364 rows

4. Loading Conflict data...
   ✓ Loaded 350 rows

5. Loading MPI (Poverty) data...
   ✓ Loaded 48 rows
✓ ALL DATASETS LOADED Sawa sawa!


### **2.3 Initial Data Exploration**

<div class="insight-box">
<p>I will preview the first few rows of each dataset to better understand the structure, columns, and sample values.</p>
</div>


#### **2.3.1 IPC Dataset Overview**

The IPC dataset provides county-level food insecurity assessments, including historical and projected phases of food insecurity as shown below.  

| <span style="color:#1F618D;">#</span> | <span style="color:#1F618D;">Column Name</span> | <span style="color:#1F618D;">Description</span> |
|----|--------------------|-------------|
| 1 | <b>Date of analysis</b> | Date when the IPC assessment was conducted and the report written |
| 2 | <b>Country</b> | Country code (always `KEN` for Kenya) |
| 3 | <b>Total country population</b> | Total population of the analyzed area (county or region) |
| 4 | <b>Level 1</b> | Administrative level name (often blank; not used in our analysis) |
| 5 | <b>Area</b> | Name of the county or region being analyzed |
| 6 | <b>Validity period</b> | Indicates whether this is a current assessment or a future projection:  

> IPC data  —  target variable. Essential. Contains county-level food insecurity phases over time.



In [14]:
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34m1. IPC DATA (Food Insecurity Levels)\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m")

# Shape
print(f"Shape: {ipc_df.shape[0]} rows × {ipc_df.shape[1]} columns\n")

# Preview first 5 rows
print("First 5 rows:")
display(ipc_df.head())

# Column names
print("\nColumn names:")
print(ipc_df.columns.tolist())


1. IPC DATA (Food Insecurity Levels)
Shape: 4522 rows × 11 columns

First 5 rows:


,Date of analysis,Country,Total country population,Level 1,Area,Validity period,From,To,Phase,Number,Percentage
0,Jul 2025,KEN,16617000,Others,Marsabit,current,2025-07-01,2025-09-30,all,515000,1.00
1,Jul 2025,KEN,16617000,Others,Marsabit,current,2025-07-01,2025-09-30,3+,103000,0.20
2,Jul 2025,KEN,16617000,Others,Marsabit,current,2025-07-01,2025-09-30,1,128750,0.25
3,Jul 2025,KEN,16617000,Others,Marsabit,current,2025-07-01,2025-09-30,2,283250,0.55
4,Jul 2025,KEN,16617000,Others,Marsabit,current,2025-07-01,2025-09-30,3,103000,0.20



Column names:
['Date of analysis', 'Country', 'Total country population', 'Level 1', 'Area', 'Validity period', 'From', 'To', 'Phase', 'Number', 'Percentage']


<blockquote>
  <p><strong>📝 Dataset Overview</strong></p>
  <p>
    The dataset contains <strong>4,522 rows</strong> and <strong>11 columns</strong> in long format. 
    Each county appears multiple times per assessment — once for each IPC phase.
  </p>
  <p>
    The <strong>IPC 3+ row</strong> (Crisis or worse) will be used to construct the target variable, 
    as it represents the population experiencing severe food insecurity.
  </p>
  <p>
    The <strong>Validity period</strong> column distinguishes current assessments from projections. 
    This distinction is critical during feature engineering to prevent data leakage and ensure 
    that only information available at the time of prediction is used.
  </p>
</blockquote>

#### **🌧2.3.2 Rainfall Data Overview**

The rainfall dataset contains dekadal (10-day period) satellite-derived rainfall measurements aggregated at administrative levels as shown below. 

| <span style="color:#1F618D;">#</span> | <span style="color:#1F618D;">Variable</span> | <span style="color:#1F618D;">Description</span> |
|----|-----------|-------------|
| 1 | <b>date</b> | Start date of the dekad (10-day period) |
| 2 | <b>adm_level</b> | Administrative level (`1` = County/Region, `2` = Sub-county) |
| 3 | <b>adm_id</b> | Unique administrative unit identification number |
| 4 | <b>PCODE</b> | Unique geographic place code (e.g., `KE019`) |
| 5 | <b>n_pixels</b> | Number of satellite pixels covering the administrative area |
| 6 | <b>rfh</b> | Observed rainfall for the dekad (millimeters) |
| 7 | <b>rfh_avg</b> | Long-term average rainfall for the same dekad (millimeters) |
| 8 | <b>r1h</b> | Cumulative 1-month rainfall (millimeters) |






In [17]:
print("\033[1;34m" + "="*80 + "\033[0m")  
print("\033[1;34m2. RAINFALL DATA\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m")

# Shape
print(f"Shape: {rainfall_df.shape[0]} rows × {rainfall_df.shape[1]} columns\n")

# Preview first 5 rows
print("First 5 rows:")
display(rainfall_df.head())

# Column names
print("\nColumn names:")
print(rainfall_df.columns.tolist())

2. RAINFALL DATA
Shape: 131544 rows × 15 columns

First 5 rows:


,date,adm_level,adm_id,PCODE,n_pixels,rfh,rfh_avg,r1h,r1h_avg,r3h,r3h_avg,rfq,r1q,r3q,version
0,1981-01-01,1,51325,KE019,427.0,7.372365,15.759407,NaN,NaN,NaN,NaN,59.598840,NaN,NaN,final
1,1981-01-11,1,51325,KE019,427.0,4.325527,19.294770,NaN,NaN,NaN,NaN,38.384920,NaN,NaN,final
2,1981-01-21,1,51325,KE019,427.0,5.569087,16.265417,17.266980,51.319595,NaN,NaN,49.700817,39.536823,NaN,final
3,1981-02-01,1,51325,KE019,427.0,5.882904,12.719282,15.777517,48.279470,NaN,NaN,61.418427,38.997230,NaN,final
4,1981-02-11,1,51325,KE019,427.0,17.180328,18.768618,28.632318,47.753320,NaN,NaN,93.317700,63.753933,NaN,final



Column names:
['date', 'adm_level', 'adm_id', 'PCODE', 'n_pixels', 'rfh', 'rfh_avg', 'r1h', 'r1h_avg', 'r3h', 'r3h_avg', 'rfq', 'r1q', 'r3q', 'version']


<blockquote>
  <p><strong>📝  Dataset Overview</strong></p>
  <p>
    The rainfall dataset contains <strong>131,544 rows</strong> and <strong>15 columns</strong>, 
    with dekadal (10-day) measurements from 1981 to the present.
  </p>
  <p>
    In addition to raw rainfall (<code>rfh</code>), the dataset includes long-term averages 
    and cumulative 1-month and 3-month totals. This structure allows it to capture both 
    short-term rainfall shocks and longer seasonal drought conditions.
  </p>
  <p>
    The ratio variables (<code>rfq</code>, <code>r1q</code>, <code>r3q</code>) — observed versus 
    historical average — are especially valuable as rainfall anomaly indicators.
  </p>
  <p>
    Some early rows contain <code>NaN</code> values in cumulative columns, which is expected 
    due to the rolling calculations used to compute those totals.
  </p>
</blockquote>

#### 🍽 **2.3.3 Food Prices Dataset Overview**

The food prices dataset provides market-level price observations for various commodities across Kenya.  
These data are essential for understanding the economic drivers of food insecurity and are used as features in the predictive model.

| <span style="color:#1F618D;">#</span> | <span style="color:#1F618D;">Column Name</span> | <span style="color:#1F618D;">Description</span> |
|----|----------------|-------------|
| 1  | <b>date</b> | Date of price observation (YYYY-MM-DD) |
| 2  | <b>admin1</b> | Administrative region (7 old regions) |
| 3  | <b>admin2</b> | Sub-region / county |
| 4  | <b>market</b> | Market location name |
| 5  | <b>market_id</b> | Unique market identifier |
| 6  | <b>latitude</b> | Market GPS latitude |
| 7  | <b>longitude</b> | Market GPS longitude |
| 8  | <b>category</b> | Food category (cereals, pulses, etc.) |
| 9  | <b>commodity</b> | Specific food item (e.g., Maize, Beans) |
| 10 | <b>commodity_id</b> | Unique commodity code |
| 11 | <b>unit</b> | Measurement unit (KG, 90 KG bag, etc.) |
| 12 | <b>priceflag</b> | Data quality flag |
| 13 | <b>pricetype</b> | Wholesale or Retail price |
| 14 | <b>currency</b> | Local currency (KES) |
| 15 | <b>price</b> | Price in local currency (KES) |
| 16 | <b>usdprice</b> | Price in US Dollars |

> Food prices  — Essential predictor. Market-level price data since 2006. 

In [20]:
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34m3. FOOD PRICES DATA\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m")

# Shape
print(f"Shape: {prices_df.shape[0]} rows × {prices_df.shape[1]} columns\n")

# Preview first 5 rows
print("First 5 rows:")
display(prices_df.head())

# Column names
print("\nColumn names:")
print(prices_df.columns.tolist())

3. FOOD PRICES DATA
Shape: 17364 rows × 16 columns

First 5 rows:


,#date,#adm1+name,#adm2+name,#loc+market+name,#loc+market+code,#geo+lat,#geo+lon,#item+type,#item+name,#item+code,#item+unit,#item+price+flag,#item+price+type,#currency+code,#value,#value+usd
0,2006-01-15,Coast,Mombasa,Mombasa,191,-4.05,39.67,cereals and tubers,Maize,51,KG,actual,Wholesale,KES,16.13,0.22
1,2006-01-15,Coast,Mombasa,Mombasa,191,-4.05,39.67,cereals and tubers,Maize (white),67,90 KG,actual,Wholesale,KES,1480.00,20.58
2,2006-01-15,Coast,Mombasa,Mombasa,191,-4.05,39.67,pulses and nuts,Beans,50,KG,actual,Wholesale,KES,33.63,0.47
3,2006-01-15,Eastern,Kitui,Kitui,187,-1.37,38.02,cereals and tubers,Maize (white),67,KG,actual,Retail,KES,17.00,0.24
4,2006-01-15,Eastern,Kitui,Kitui,187,-1.37,38.02,cereals and tubers,Sorghum,65,90 KG,actual,Wholesale,KES,1800.00,25.03



Column names:
['#date', '#adm1+name', '#adm2+name', '#loc+market+name', '#loc+market+code', '#geo+lat', '#geo+lon', '#item+type', '#item+name', '#item+code', '#item+unit', '#item+price+flag', '#item+price+type', '#currency+code', '#value', '#value+usd']


<blockquote>
  <p><strong>📝  Dataset Overview</strong></p>
  <p>
    The food prices dataset contains <strong>17,364 rows</strong> and <strong>16 columns</strong>, 
    covering market-level commodity prices across Kenya from 2006 onward.
  </p>
  <p>
    It includes both wholesale and retail prices for key staples such as maize, beans, 
    and sorghum across multiple counties.
  </p>
  <p>
    Column names follow a Humanitarian Data Exchange (HDX) naming convention in the raw file 
    and will need to be standardized during preprocessing for consistency.
  </p>
  <p>
    Prices are reported in both KES and USD, allowing flexibility for currency conversion 
    and inflation-adjusted analysis.
  </p>
  
</blockquote>

#### **2.3.4 Conflict Dataset Overview**

The conflict dataset provides monthly records of political violence events and fatalities in Kenya.  

| # | Column Name | Description |
|----|----------------|-------------|
| 1  | **Country** | Country name (Kenya) |
| 2  | **Month** | Month of observation |
| 3  | **Year** | Year of observation |
| 4  | **Events** | Number of recorded violent events |
| 5  | **Fatalities** | Number of deaths resulting from violence |




In [23]:
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34m4. CONFLICT DATA (Violence Events)\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m")

# Shape
print(f"Shape: {conflict_df.shape[0]} rows × {conflict_df.shape[1]} columns\n")

# Preview first 5 rows
print("First 5 rows:")
display(conflict_df.head())

# Column names
print("\nColumn names:")
print(conflict_df.columns.tolist())

4. CONFLICT DATA (Violence Events)
Shape: 350 rows × 5 columns

First 5 rows:


,Country,Month,Year,Events,Fatalities
0,Kenya,January,1997,3,6
1,Kenya,February,1997,3,9
2,Kenya,March,1997,6,179
3,Kenya,April,1997,4,7
4,Kenya,May,1997,4,22



Column names:
['Country', 'Month', 'Year', 'Events', 'Fatalities']


<blockquote>
  <p><strong>📝 Dataset Overview</strong></p>
  <p>
    The conflict dataset contains <strong>350 rows</strong> and <strong>5 columns</strong>, 
    covering monthly violence events and fatalities in Kenya from 1997 onward.
  </p>
  <p>
    It is the most compact of all datasets and operates at the <strong>national level</strong> 
    rather than the county level. As a result, it captures overall instability trends 
    rather than localized conflict dynamics.
  </p>
  <p>
    While this limits spatial granularity, the dataset remains a valuable contextual 
    feature  particularly during periods of election-related violence or broader 
    nationwide insecurity.
  </p>
</blockquote>

#### **2.3.5 MPI Dataset Overview (Poverty Levels)**

The MPI dataset provides county-level Multidimensional Poverty Index (MPI) indicators for Kenya.  
These data measure structural poverty and long-term deprivation across counties.  
They are used as background features in the predictive model to capture underlying vulnerability.

| # | Column Name | Description |
|----|-------------------------------|-------------|
| 1  | **#country+code** | Country code (KEN) |
| 2  | **#adm1+code** | County code (e.g., KE001) |
| 3  | **#adm1+name** | County name |
| 4  | **#indicator+mpi** | Multidimensional Poverty Index value |
| 5  | **#indicator+headcount_ratio** | Percentage of population classified as multidimensionally poor |
| 6  | **#indicator+intensity_of_deprivation** | Average deprivation intensity among the poor |
| 7  | **#indicator+vulnerable_to_poverty** | Percentage vulnerable to falling into poverty |
| 8  | **#indicator+in_severe_poverty** | Percentage living in severe poverty |
| 9  | **#meta+survey** | Survey source (DHS) |
| 10 | **#date+start** | Survey start date |
| 11 | **#date+end** | Survey end date |



In [26]:
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34m5. MPI DATA (Poverty Levels)\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m")

# Shape
print(f"Shape: {mpi_df.shape[0]} rows × {mpi_df.shape[1]} columns\n")

# Preview first 5 rows
print("First 5 rows:")
display(mpi_df.head())

# Column names
print("\nColumn names:")
print(mpi_df.columns.tolist())


5. MPI DATA (Poverty Levels)
Shape: 48 rows × 11 columns

First 5 rows:


,#country+code,#adm1+code,#adm1+name,#indicator+mpi,#indicator+headcount_ratio,#indicator+intensity_of_deprivation,#indicator+vulnerable_to_poverty,#indicator+in_severe_poverty,#meta+survey,#date+start,#date+end
0,KEN,NaN,NaN,0.1134,25.3523,44.7108,26.4044,7.4594,DHS,2022-01-01 00:00:00+00:00,2022-12-31 23:59:59+00:00
1,KEN,KE001,Mombasa,0.0518,12.8866,40.2193,16.8930,2.2801,DHS,2022-01-01 00:00:00+00:00,2022-12-31 23:59:59+00:00
2,KEN,KE002,Kwale,0.2105,44.5123,47.2996,27.2794,17.2902,DHS,2022-01-01 00:00:00+00:00,2022-12-31 23:59:59+00:00
3,KEN,KE003,Kilifi,0.2026,46.3581,43.7119,23.8772,13.1869,DHS,2022-01-01 00:00:00+00:00,2022-12-31 23:59:59+00:00
4,KEN,KE004,Tana River,0.3780,67.2861,56.1740,18.3876,44.4231,DHS,2022-01-01 00:00:00+00:00,2022-12-31 23:59:59+00:00



Column names:
['#country+code', '#adm1+code', '#adm1+name', '#indicator+mpi', '#indicator+headcount_ratio', '#indicator+intensity_of_deprivation', '#indicator+vulnerable_to_poverty', '#indicator+in_severe_poverty', '#meta+survey', '#date+start', '#date+end']


<blockquote>
  <p><strong>📝 Dataset Overview</strong></p>
  <p>
    The MPI dataset contains <strong>48 rows</strong> and <strong>11 columns</strong>, 
    with one row per county plus a national aggregate (row 0, where <code>#adm1+code</code> is NaN).
  </p>
  <p>
    Unlike the other datasets, this is <strong>static cross-sectional data</strong> 
    from the 2022 DHS survey. It will therefore be merged as a fixed structural feature 
    rather than a time-varying predictor.
  </p>
  <p>
    Counties such as <strong>Tana River (MPI: 0.378)</strong> and 
    <strong>Kwale (MPI: 0.210)</strong> already exhibit high structural poverty, 
    which is expected to correlate strongly with food insecurity risk.
  </p>
  <p>
    Column names follow the HDX naming convention and will need to be standardized 
    during preprocessing.
  </p>
</blockquote>

### **2.4 Data Quality Check**



#### **2.4.1 IPC Data Types & Missing Values**


In [30]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34m1. IPC DATA - Info and Missing Values\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Data types
print("\033[1;33mData Types:\033[0m")  # Yellow header
print(ipc_df.dtypes)

# Missing values
missing_ipc = ipc_df.isnull().sum()
print("\n\033[1;33mMissing Values:\033[0m")
if missing_ipc.sum() == 0:
    print("✓ No missing values!")
else:
    print(missing_ipc[missing_ipc > 0])

# Missing value percentage
total_missing = ipc_df.isnull().sum().sum()
total_cells = ipc_df.shape[0] * ipc_df.shape[1]
missing_percentage = (total_missing / total_cells) * 100
print(f"\nMissing percentage: {missing_percentage:.2f}%")

1. IPC DATA - Info and Missing Values

Data Types:
Date of analysis                str
Country                         str
Total country population      int64
Level 1                         str
Area                            str
Validity period                 str
From                            str
To                              str
Phase                           str
Number                        int64
Percentage                  float64
dtype: object

Missing Values:
Level 1    1610
dtype: int64

Missing percentage: 3.24%


<blockquote>
  <p><strong>📝 findingsk</strong></p>
  <p>
    The IPC dataset has a very low missing rate of <strong>3.24%</strong>, entirely 
    concentrated in the <code>Level 1</code> column (1,610 missing values).
  </p>
  <p>
    This is not a concern, as <code>Level 1</code> was previously identified as 
    unused in the analysis. All other columns are complete.
  </p>
  <p>
    The date-related fields (<code>From</code>, <code>To</code>, and 
    <code>Date of analysis</code>) are currently stored as strings and will 
    need to be converted to <code>datetime</code> format during preprocessing.
  </p>
</blockquote>

#### **2.4.2 Rainfall Data Types & Missing Values**

In [33]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34m2. RAINFALL DATA - Info and Missing Values\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Data types
print("\033[1;33mData Types:\033[0m")  # Yellow header
print(rainfall_df.dtypes)

# Missing values
missing_rain = rainfall_df.isnull().sum()
print("\n\033[1;33mMissing Values:\033[0m")
if missing_rain.sum() == 0:
    print("✓ No missing values!")
else:
    print(missing_rain[missing_rain > 0])

# Missing value percentage
total_missing = rainfall_df.isnull().sum().sum()
total_cells = rainfall_df.shape[0] * rainfall_df.shape[1]
missing_percentage = (total_missing / total_cells) * 100
print(f"\nMissing percentage: {missing_percentage:.2f}%")

2. RAINFALL DATA - Info and Missing Values

Data Types:
date             str
adm_level      int64
adm_id         int64
PCODE            str
n_pixels     float64
rfh          float64
rfh_avg      float64
r1h          float64
r1h_avg      float64
r3h          float64
r3h_avg      float64
rfq          float64
r1q          float64
r3q          float64
version          str
dtype: object

Missing Values:
r1h        162
r1h_avg    162
r3h        648
r3h_avg    648
r1q        162
r3q        648
dtype: int64

Missing percentage: 0.12%


<blockquote>
  <p><strong>📝 Findings</strong></p>
  <p>
    The rainfall dataset is highly complete, with a missing rate of just 
    <strong>0.12%</strong>.
  </p>
  <p>
    Missing values are confined to cumulative columns (<code>r1h</code>, 
    <code>r3h</code>, and their corresponding averages and ratios). This is expected, 
    as rolling aggregations naturally produce <code>NaN</code> values at the beginning 
    of each administrative unit’s time series before sufficient dekads accumulate.
  </p>
  <p>
    These initial gaps will resolve during feature engineering and do not require 
    imputation.
  </p>
  <p>
    The <code>date</code> column is currently stored as a string and will need to be 
    converted to <code>datetime</code> format during preprocessing.
  </p>
</blockquote>

#### **2.4.3 Food prices Data Types & Missing Values**

In [36]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m")  
print("\033[1;34m3. FOOD PRICES DATA - Info and Missing Values\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Data types
print("\033[1;33mData Types:\033[0m")  
print(prices_df.dtypes)

# Missing values
missing_prices = prices_df.isnull().sum()
print("\n\033[1;33mMissing Values:\033[0m")
if missing_prices.sum() == 0:
    print("✓ No missing values!")
else:
    print(missing_prices[missing_prices > 0])

# Missing value percentage
total_missing = prices_df.isnull().sum().sum()
total_cells = prices_df.shape[0] * prices_df.shape[1]
missing_percentage = (total_missing / total_cells) * 100
print(f"\nMissing percentage: {missing_percentage:.2f}%")

3. FOOD PRICES DATA - Info and Missing Values

Data Types:
#date                   str
#adm1+name              str
#adm2+name              str
#loc+market+name        str
#loc+market+code      int64
#geo+lat            float64
#geo+lon            float64
#item+type              str
#item+name              str
#item+code            int64
#item+unit              str
#item+price+flag        str
#item+price+type        str
#currency+code          str
#value              float64
#value+usd          float64
dtype: object

Missing Values:
#adm1+name    50
#adm2+name    50
#geo+lat      50
#geo+lon      50
dtype: int64

Missing percentage: 0.07%


<blockquote>
  <p><strong>📝 Findings</strong></p>
  <p>
    The food prices dataset is nearly complete, with a very low missing rate of 
    <strong>0.07%</strong>.
  </p>
  <p>
    The 50 missing values across <code>#adm1+name</code>, <code>#adm2+name</code>, 
    <code>#geo+lat</code>, and <code>#geo+lon</code> occur together, indicating they 
    belong to the same 50 records with unidentified market locations.
  </p>
  <p>
    Since county-level aggregation requires a valid <code>#adm2+name</code>, these 
    records will need to be either dropped or further investigated during preprocessing.
  </p>
  <p>
    All other columns are complete. The <code>#date</code> column is currently stored 
    as a string and will require conversion to <code>datetime</code> format.
  </p>
</blockquote>

#### **2.4.4 Conflict Data Types & Missing Values**

In [39]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m") 
print("\033[1;34m4. CONFLICT DATA - Info and Missing Values\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Data types
print("\033[1;33mData Types:\033[0m") 
print(conflict_df.dtypes)

# Missing values
missing_conflict = conflict_df.isnull().sum()
print("\n\033[1;33mMissing Values:\033[0m")
if missing_conflict.sum() == 0:
    print("✓ No missing values!")
else:
    print(missing_conflict[missing_conflict > 0])

# Missing value percentage
total_missing = conflict_df.isnull().sum().sum()
total_cells = conflict_df.shape[0] * conflict_df.shape[1]
missing_percentage = (total_missing / total_cells) * 100
print(f"\nMissing percentage: {missing_percentage:.2f}%")

4. CONFLICT DATA - Info and Missing Values

Data Types:
Country         str
Month           str
Year          int64
Events        int64
Fatalities    int64
dtype: object

Missing Values:
✓ No missing values!

Missing percentage: 0.00%


<blockquote>
  <p><strong>📝Findings</strong></p>
  <p>
    The conflict dataset is fully complete, with <strong>no missing values</strong>.
  </p>
  <p>
    The only preprocessing required is to combine the <code>Month</code> and 
    <code>Year</code> string columns into a single <code>datetime</code> field. 
    This will allow proper temporal alignment when merging with the other datasets.
  </p>
</blockquote>

#### **2.4.5 MPI Data Types & Missing Values**

In [42]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34m5. MPI DATA - Info and Missing Values\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Data types
print("\033[1;33mData Types:\033[0m")  # Yellow header
print(mpi_df.dtypes)

# Missing values
missing_mpi = mpi_df.isnull().sum()
print("\n\033[1;33mMissing Values:\033[0m")
if missing_mpi.sum() == 0:
    print("✓ No missing values!")
else:
    print(missing_mpi[missing_mpi > 0])

# Missing value percentage
total_missing = mpi_df.isnull().sum().sum()
total_cells = mpi_df.shape[0] * mpi_df.shape[1]
missing_percentage = (total_missing / total_cells) * 100
print(f"\nMissing percentage: {missing_percentage:.2f}%")

5. MPI DATA - Info and Missing Values

Data Types:
#country+code                              str
#adm1+code                                 str
#adm1+name                                 str
#indicator+mpi                         float64
#indicator+headcount_ratio             float64
#indicator+intensity_of_deprivation    float64
#indicator+vulnerable_to_poverty       float64
#indicator+in_severe_poverty           float64
#meta+survey                               str
#date+start                                str
#date+end                                  str
dtype: object

Missing Values:
#adm1+code    1
#adm1+name    1
dtype: int64

Missing percentage: 0.38%


<blockquote>
  <p><strong>📝 Findings</strong></p>
  <p>
    The MPI dataset is nearly complete, with a very low missing rate of 
    <strong>0.38%</strong>.
  </p>
  <p>
    The single missing entry in <code>#adm1+code</code> and <code>#adm1+name</code> 
    corresponds to the national aggregate record (row 0) identified earlier. 
    This row will be excluded during county-level merging.
  </p>
  <p>
    All poverty indicator columns are fully complete and ready to be used 
    as static structural features.
  </p>
</blockquote>

  <p>
    <strong>Summary:</strong> Core predictive features will come from IPC, rainfall, and food prices. MPI provides structural context, while conflict may add limited value during periods of national unrest.
  </p>

### **2.5 Basic Statistics**

#### **2.5.1 IPC Data**

In [47]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34mIPC DATA - Statistical Summary\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Display descriptive statistics
display(ipc_df.describe())

IPC DATA - Statistical Summary



,Total country population,Number,Percentage
count,4.522000e+03,4.522000e+03,4522.000000
mean,1.603694e+07,1.987975e+05,0.306347
std,1.079520e+06,2.899369e+05,0.344474
min,1.388144e+07,0.000000e+00,0.000000
25%,1.515218e+07,0.000000e+00,0.000000
50%,1.661700e+07,7.885000e+04,0.150000
75%,1.661700e+07,2.680020e+05,0.450000
max,1.790982e+07,1.577335e+06,1.000000


<blockquote>
  <p><strong>📝 Key insights</strong></p>
  <ul>
    <li><strong>Total:</strong> Country population ranges from ~13.9M to ~17.9M across assessment years, reflecting overall population growth rather than county-level differences.</li>
    <li><strong>Number:</strong> Population count per IPC phase averages ~199K, with a high standard deviation (~290K). This indicates substantial variation across counties and phases. Some rows record zero where a phase is absent in a county.</li>
    <li><strong>Percentage:</strong> Averages 0.31, with the 25th percentile at 0, showing many county-phase combinations have no population in that phase. The maximum of 1.0 occurs where the entire assessed population is in a single phase.</li>
  </ul>
 
</blockquote>

#### **2.5.2 Rainfall Data**

In [50]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34mRAINFALL DATA - Statistical Summary\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Display descriptive statistics
display(rainfall_df.describe())

RAINFALL DATA - Statistical Summary



,adm_level,adm_id,n_pixels,rfh,rfh_avg,r1h,r1h_avg,r3h,r3h_avg,rfq,r1q,r3q
count,131544.000000,131544.000000,131544.000000,131544.000000,131544.000000,131382.000000,131382.000000,130896.000000,130896.000000,131544.000000,131382.000000,130896.000000
mean,1.901235,52031.049383,457.962963,31.660675,30.945180,95.053785,92.923584,285.738097,278.942779,101.535986,101.622794,102.174143
std,0.298348,3054.654585,1007.072725,33.554252,22.425579,82.330986,63.801213,186.729623,153.399693,58.847724,47.449597,36.749438
min,1.000000,51325.000000,10.000000,0.000000,0.000000,0.175904,0.449237,2.139759,3.654337,9.762118,10.142180,15.911969
25%,2.000000,51345.000000,31.000000,7.173302,13.146297,30.231770,43.208694,138.574680,159.248900,62.904640,70.349533,78.376366
50%,2.000000,51365.000000,70.000000,20.421646,27.449320,74.578945,84.123810,259.145325,268.148070,90.170180,93.945730,97.291840
75%,2.000000,51385.000000,311.000000,45.090908,44.684128,138.706545,133.981600,403.485965,388.407300,121.772941,122.011882,118.613502
max,2.000000,68770.000000,5604.000000,476.275850,131.643330,975.482800,332.292140,1495.413700,740.640200,770.968750,571.586730,460.272060


<blockquote>
  <p><strong>📝 Key insights</strong></p>
  <ul>
    <li><strong>Mean dekadal rainfall (rfh):</strong> 31.7mm, closely matching the long-term average (rfh_avg) of 30.9mm, indicating data aligns with historical trends.</li>
    <li><strong>Variability:</strong> High standard deviation (~33.5mm) and maximum of 476mm reveal substantial differences across counties and seasons.</li>
    <li><strong>Anomaly ratios (rfq, r1q, r3q):</strong> Average ~100-102, indexed to long-term averages. Values below 100 indicate below-average rainfall and serve as key drought signals.</li>
    <li><strong>n_pixels:</strong> Ranges from 10 to 5,604, reflecting differences in county/sub-county size.</li>
    <li><strong>adm_level:</strong> Mostly 2, meaning the dataset is predominantly at the sub-county level. Aggregation to county level will be needed before merging.</li>
  </ul>
  
</blockquote>

#### **2.5.3 Food Prices Data**

In [53]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34mFOOD PRICES DATA - Statistical Summary\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Display descriptive statistics
display(prices_df.describe())

FOOD PRICES DATA - Statistical Summary



,#loc+market+code,#geo+lat,#geo+lon,#item+code,#value,#value+usd
count,17364.000000,17314.000000,17314.000000,17364.000000,17364.000000,17364.000000
mean,3968.962797,0.677603,36.973523,291.947996,1369.549790,12.592478
std,3171.609080,2.285165,2.011423,294.003509,2606.960335,23.259702
min,184.000000,-4.660000,34.360000,50.000000,5.000000,0.039000
25%,1850.000000,-1.270000,34.860000,67.000000,67.500000,0.600000
50%,3254.000000,0.190000,36.790000,148.000000,136.000000,1.060000
75%,5670.000000,3.240000,38.660000,440.000000,1389.160000,13.560000
max,10509.000000,4.620000,41.860000,937.000000,18252.000000,150.950000


#### **2.5.4 Conflict Data**

In [55]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34mCONFLICT DATA - Statistical Summary\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Display descriptive statistics
display(conflict_df.describe())

CONFLICT DATA - Statistical Summary



,Year,Events,Fatalities
count,350.000000,350.000000,350.000000
mean,2011.085714,27.702857,41.554286
std,8.430982,32.054802,47.350261
min,1997.000000,0.000000,0.000000
25%,2004.000000,8.000000,12.000000
50%,2011.000000,15.000000,29.000000
75%,2018.000000,29.000000,58.750000
max,2026.000000,175.000000,539.000000


#### **2.5.5 MPI Data**

In [57]:
# Styled header
print("\033[1;34m" + "="*80 + "\033[0m")  # Blue header
print("\033[1;34mMPI DATA - Statistical Summary\033[0m")
print("\033[1;34m" + "="*80 + "\033[0m\n")

# Display descriptive statistics
display(mpi_df.describe())

MPI DATA - Statistical Summary



,#indicator+mpi,#indicator+headcount_ratio,#indicator+intensity_of_deprivation,#indicator+vulnerable_to_poverty,#indicator+in_severe_poverty
count,48.000000,48.000000,48.000000,48.00000,48.000000
mean,0.152240,32.107352,43.979956,26.99845,12.248000
std,0.120417,19.724275,6.251215,9.21059,16.516930
min,0.010300,2.671200,37.766800,9.21130,0.221300
25%,0.078500,19.149225,40.000075,18.34410,2.342800
50%,0.108250,25.707150,40.998600,27.20630,5.313800
75%,0.194100,40.455875,45.955325,33.44040,12.004975
max,0.496800,81.335900,62.447900,43.37600,64.885500


### **2.6 Unique Values in Key Columns**